Uncouple replay creation and rendering

gberaudo opened the pull request:
The goal of this PR is to split the replays in 2 parts:
- classes for building instructions (typically used in web workers);
- classes for executing the instructions (typically used on the main thread).

Plan is:
- [x] have a replay group instance for building instructions and one for executing them;
- [x] be able to extract and replace instructions;
- [x] physically split the builders and executor in separate classes;
- [ ] only pass information in instructions and coordinate array;

Postponed to a later phase:
- [ ] make instructions serializable (remove image creation from the styles, create and cache images in the instruction executor);
- [ ] binary transferable representation of the instructions (will be a lot of work)

fredj commented on src/ol/render/canvas/Replay.js:
```diff
@@ -28,6 +28,14 @@ import {
+ * @property {Array<*>} instructions The rendering instructions
+ * @property {Array<*>} hitDetectionInstructions The rendering hit detection instructions
+ * @property {Array<number>} coordinates
+ * }}
```
extra `}}`

gberaudo replied:
Thanks

fredj commented on src/ol/render/canvas/Replay.js:
```diff
@@ -28,6 +28,16 @@ import {
+ * @property {Array<number>} coordinates The array of all coordinates.
+ * @property {!Object<string, import("../canvas.js").TextState>} textStates The text states (decluttering).
+ * @property {!Object<string, import("../canvas.js").FillState>} fillStates The fill states (decluttering).
+ * @property {!Object<string, import("../canvas.js").StrokeState>} strokeStates The stoke states (decluttering).
```
`stroke`instead of `stoke`

ahocevar commented on src/ol/render/canvas/InstructionsExecutor.js:
```diff
@@ -0,0 +1,1369 @@
+  /**
+   * @inheritDoc
+   */
+  drawText(geometry, feature) {
```
This function is not used here.

ahocevar replied:
I removed it.

ahocevar commented on src/ol/render/canvas/InstructionsExecutor.js:
```diff
@@ -0,0 +1,1369 @@
+/**
+ * @module ol/render/canvas/Replay
```
should be `@module ol/render/canvas/InstructionsExecutor`

fredj commented on src/ol/render/canvas/Replay.js:
```diff
@@ -456,9 +476,15 @@ class CanvasReplay extends VectorContext {
 
   /**
-   * FIXME empty description for jsdoc
+   * @return {Object} the serializable instructions.
```
Change `Object` to `SerializableInstructions`, and maybe return `undefined`values for `textStates`, `fillStates` and `strokeStates` ?

fredj commented on src/ol/render/canvas/InstructionsGroupExecutor.js:
```diff
@@ -0,0 +1,491 @@
+/**
+ * @module ol/render/canvas/ReplayGroup
```
`@module ol/render/canvas/InstructionsGroupExecutor`

gberaudo commented:
@ahocevar, the creation of text instructions is now fully separated from the executor.
I will do some cosmetics cleanups and after that it should be in good shape for merging.

I will work on removing the creation of images from the styles in a different PR as it is not necessary until we have workers and I am eager to see this PR merged.

gberaudo commented:
@ahocevar, the PR is now ready for review.

gberaudo commented:
@ahocevar, I fixed the pixelratio issue, improved the test and fixed the mismatching executor/builder types.

I also simplified the names: Executor, Builder, ExecutorGroup, BuilderGroup. That is shorter to type and easier to remember.

The last point left is passing all states in the instructions which I will tackle tomorrow.
